_posts/2024-01-16-fido-passkey.md

```diff
@@ -0,0 +1,1127 @@
+        }
+
+    }
+
+    private User getUser(String username) {
+        Optional<User> user = userRepository.findByUsername(username);
+        if (user.isEmpty() || !user.get().isRegistrationComplete()) {
+            if (user.isEmpty()) {
+                log.error("User with username {} not found", username);
+            } else {
+                log.error("Registration for username {} is not complete", username);
+            }
+            throw new RuntimeException(String.format("User with username %s not registered", username));
+        }
+        return user.get();
+    }
+
+    private AssertionRequest createAssertionRequest(String json) throws JsonProcessingException {
+        return AssertionRequest.fromJson(json);
+    }
+
+    private AuthenticatorAssertionResponse createAuthenticatorAssertionResponse(VerifyLoginRequestResource resource) throws Base64UrlException, IOException {
+        return AuthenticatorAssertionResponse.builder()
+                .authenticatorData(resource.response().response().authenticatorData())
+                .clientDataJSON(resource.response().response().clientDataJSON())
+                .signature(resource.response().response().signature())
+                .build();
+    }
+
+    private PublicKeyCredential createPublicKeyCredential(VerifyLoginRequestResource resource, AuthenticatorAssertionResponse response) {
+        CustomClientExtensionOutput customClientExtensionOutput = new CustomClientExtensionOutput();
+
+        return PublicKeyCredential.builder()
+                .id(resource.response().id())
+                .response(response)
+                .clientExtensionResults(customClientExtensionOutput)
+                .build();
+    }
+
+    private FinishAssertionOptions createFinishAssertionOptions(AssertionRequest assertionRequest, PublicKeyCredential publicKeyCredential) {
+        return FinishAssertionOptions.builder()
+                .request(assertionRequest)
+                .response(publicKeyCredential)
+                .build();
+    }
+}
+```
+
+### Login Controller
+Create a controller `LoginController` with 2 endpoints, the `start login` and `verify login`.
+
+```java
+@AllArgsConstructor
+@RestController
+@RequestMapping("/login")
+@CrossOrigin("http://localhost:4200")
+public class LoginController {
+    private final StartLoginService startLoginService;
+    private final VerifyLoginService verifyLoginService;
+
+    @PostMapping("/start")
+    public ResponseEntity<LoginResponseResource> startLogin(@RequestBody LoginRequestResource resource) throws JsonProcessingException {
+        return ResponseEntity.ok(startLoginService.startLogin(resource));
+    }
+
+    @PostMapping("/verify")
+    public ResponseEntity<VerifyLoginResponseResource> verifyLogin(@RequestBody VerifyLoginRequestResource resource) throws Base64UrlException, IOException {
+        return ResponseEntity.ok(verifyLoginService.verify(resource));
+    }
+}
+```
+
```

Comment:
Since there is very little explanation about the code itself (specific lines), consider moving it out of the blogpost (github repo) and just reference (in a bullet list or small paragraphs) what the different components will do.